## **The frist several cells are for Google Colab only**
It is used to link to your folder on Google Drive; To run the following command, remove the hashtages at teh beginning of the command lines 

In [ ]:

from google.colab import drive
drive.mount('/content/drive')

Change the directory to your folder on Google Drive

In [ ]:
cd '/content/drive/My Drive/Colab Notebooks/MKTG6010'

In [ ]:
## display the directory
!pwd

# Part 1: Data reading and preprocessing
## **Starting from now, the codes are for both Jupyter and Google Colab**

In [2]:
import pandas as pd
import numpy as np

import warnings
warnings.filterwarnings('ignore')

# Read the data
data=pd.read_csv('C:/Users/Furqan Ali/Desktop/woek/linkedin_data.csv')

# show the top 5 rows
data.head()

IPAddress  Duration (in seconds)         ResponseId  \
0  198.182.157.254                    239  R_bqkat7vEJIOtO7j   
1   174.155.134.32                    191  R_4UyBdECON03cUrb   
2     45.46.112.52                    140  R_9BL6eXmPRlTawEl   
3    172.56.23.132                    410  R_6EasCVITCdtBRI1   
4    65.96.122.144                    171  R_3PMRpp5cpHj5e1D   

   LocationLatitude  LocationLongitude   Q4  Q5_1  Q5_2  Q5_3  Q5_4  ...  \
0         34.417999        -118.566200  Yes     3     4     4     3  ...   
1         25.857605         -80.278099  Yes     4     5     4     4  ...   
2         46.912094         -68.043098  Yes     4     6     4     3  ...   
3         42.464203         -70.971703  Yes     4     4     3     4  ...   
4         43.644196         -72.306602  Yes     4     5     4     4  ...   

   Q8_14  Q8_15               Q10                Q11                 Q12  \
0    4.0    4.0  2-3 times a week          1-3 years           1-2 hours   
1    4.0    5.0  2-3 times a week        7-12 months           1-2 hours   
2    5.0    5.0  2-3 times a week          1-3 years  Less than one hour   
3    3.0    4.0  2-3 times a week        7-12 months           2-5 hours   
4    5.0    5.0  2-3 times a week  More than 3 years  Less than one hour   

      Q14    Q15                    Q16                       Q17  \
0  Female  26-35         Masters Degree  United States of America   
1  Female  26-35            High School  United States of America   
2    Male  26-35  4-year College Degree  United States of America   
3    Male  36-50            High School  United States of America   
4  Female  36-50         Masters Degree  United States of America   

                                              Q18  
0                            Educational services  
1               Real estate or rental and leasing  
2               Arts, entertainment or recreation  
3  Professional, scientific or technical services  
4                                    Construction  

[5 rows x 65 columns]

In [3]:
# Show all of the data
# We can find we have 282 rows × 65 columns
data

IPAddress  Duration (in seconds)         ResponseId  \
0    198.182.157.254                    239  R_bqkat7vEJIOtO7j   
1     174.155.134.32                    191  R_4UyBdECON03cUrb   
2       45.46.112.52                    140  R_9BL6eXmPRlTawEl   
3      172.56.23.132                    410  R_6EasCVITCdtBRI1   
4      65.96.122.144                    171  R_3PMRpp5cpHj5e1D   
..               ...                    ...                ...   
277   203.13.128.104                  48799  R_5oKT2UHrPLLN5yV   
278    24.20.159.132                    308  R_b13YF7M1qu4RycJ   
279   70.124.152.197                   1866  R_2sizYJgiDIpBdLT   
280     71.13.225.11                    291  R_9Z8cqlkP0kY5Zgp   
281     73.31.223.54                   2831  R_7OJwqZ6sI9wQsMR   

     LocationLatitude  LocationLongitude   Q4  Q5_1  Q5_2  Q5_3  Q5_4  ...  \
0           34.417999        -118.566200  Yes     3     4     4     3  ...   
1           25.857605         -80.278099  Yes     4     5     4     4  ...   
2           46.912094         -68.043098  Yes     4     6     4     3  ...   
3           42.464203         -70.971703  Yes     4     4     3     4  ...   
4           43.644196         -72.306602  Yes     4     5     4     4  ...   
..                ...                ...  ...   ...   ...   ...   ...  ...   
277        -33.791199         151.129791  Yes     7     3     3     2  ...   
278         45.514404        -122.643295  Yes     7     4     4     4  ...   
279         30.365997         -97.643700  Yes     7     5     4     4  ...   
280         42.560196         -85.858597  Yes     7     7     6     7  ...   
281               NaN                NaN  Yes     7     7     7     7  ...   

     Q8_14  Q8_15                      Q10                Q11  \
0      4.0    4.0         2-3 times a week          1-3 years   
1      4.0    5.0         2-3 times a week        7-12 months   
2      5.0    5.0         2-3 times a week          1-3 years   
3      3.0    4.0         2-3 times a week        7-12 months   
4      5.0    5.0         2-3 times a week  More than 3 years   
..     ...    ...                      ...                ...   
277    3.0    1.0  Once in a month or less  More than 3 years   
278    4.0    4.0  Once in a month or less  More than 3 years   
279    2.0    2.0  Once in a month or less          1-3 years   
280    4.0    4.0  Once in a month or less  More than 3 years   
281    NaN    NaN                      NaN                NaN   

                    Q12     Q14    Q15                    Q16  \
0             1-2 hours  Female  26-35         Masters Degree   
1             1-2 hours  Female  26-35            High School   
2    Less than one hour    Male  26-35  4-year College Degree   
3             2-5 hours    Male  36-50            High School   
4    Less than one hour  Female  36-50         Masters Degree   
..                  ...     ...    ...                    ...   
277  Less than one hour  Female  26-35    Professional Degree   
278  Less than one hour  Female  36-50  4-year College Degree   
279  Less than one hour  Female  36-50        Doctoral Degree   
280  Less than one hour  Female  36-50  4-year College Degree   
281                 NaN     NaN    NaN                    NaN   

                          Q17                                             Q18  
0    United States of America                            Educational services  
1    United States of America               Real estate or rental and leasing  
2    United States of America               Arts, entertainment or recreation  
3    United States of America  Professional, scientific or technical services  
4    United States of America                                    Construction  
..                        ...                                             ...  
277                 Australia  Professional, scientific or technical services  
278  United States of America  Professional, scientific or technical services  

In [4]:
# show the last 5 rows of the data
data.tail()

IPAddress  Duration (in seconds)         ResponseId  \
277  203.13.128.104                  48799  R_5oKT2UHrPLLN5yV   
278   24.20.159.132                    308  R_b13YF7M1qu4RycJ   
279  70.124.152.197                   1866  R_2sizYJgiDIpBdLT   
280    71.13.225.11                    291  R_9Z8cqlkP0kY5Zgp   
281    73.31.223.54                   2831  R_7OJwqZ6sI9wQsMR   

     LocationLatitude  LocationLongitude   Q4  Q5_1  Q5_2  Q5_3  Q5_4  ...  \
277        -33.791199         151.129791  Yes     7     3     3     2  ...   
278         45.514404        -122.643295  Yes     7     4     4     4  ...   
279         30.365997         -97.643700  Yes     7     5     4     4  ...   
280         42.560196         -85.858597  Yes     7     7     6     7  ...   
281               NaN                NaN  Yes     7     7     7     7  ...   

     Q8_14  Q8_15                      Q10                Q11  \
277    3.0    1.0  Once in a month or less  More than 3 years   
278    4.0    4.0  Once in a month or less  More than 3 years   
279    2.0    2.0  Once in a month or less          1-3 years   
280    4.0    4.0  Once in a month or less  More than 3 years   
281    NaN    NaN                      NaN                NaN   

                    Q12     Q14    Q15                    Q16  \
277  Less than one hour  Female  26-35    Professional Degree   
278  Less than one hour  Female  36-50  4-year College Degree   
279  Less than one hour  Female  36-50        Doctoral Degree   
280  Less than one hour  Female  36-50  4-year College Degree   
281                 NaN     NaN    NaN                    NaN   

                          Q17                                             Q18  
277                 Australia  Professional, scientific or technical services  
278  United States of America  Professional, scientific or technical services  
279  United States of America                            Educational services  
280  United States of America               Real estate or rental and leasing  
281                       NaN                                             NaN  

[5 rows x 65 columns]

In [5]:
# Create a dummy variable/column in data called 'stay' : 
# 1 - definately stay: 7
# 0 - otherwise: 1 - 6
data['stay']=(data['Q5_1']>6).astype(int)

In [6]:
# Assign X as Q5_2 : Q5_9
X=data[data.columns[7:15]]
y=data['stay']

In [7]:
#show descriptive stats of the variable 'X'
X.describe()

Q5_2        Q5_3        Q5_4        Q5_5        Q5_6        Q5_7  \
count  282.000000  282.000000  282.000000  282.000000  282.000000  282.000000   
mean     5.648936    5.056738    5.425532    4.985816    5.326241    5.209220   
std      1.225826    1.538466    1.353703    1.779623    1.622776    1.558647   
min      2.000000    1.000000    1.000000    1.000000    1.000000    1.000000   
25%      5.000000    4.000000    4.000000    4.000000    4.000000    4.000000   
50%      6.000000    5.000000    6.000000    5.000000    6.000000    5.000000   
75%      7.000000    6.000000    7.000000    7.000000    7.000000    6.000000   
max      7.000000    7.000000    7.000000    7.000000    7.000000    7.000000   

             Q5_8        Q5_9  
count  282.000000  282.000000  
mean     5.556738    3.620567  
std      1.270515    2.154529  
min      1.000000    1.000000  
25%      5.000000    1.000000  
50%      6.000000    3.500000  
75%      7.000000    5.000000  
max      7.000000    7.000000

In [8]:
#show the shape of X
# For X, we have 282 rows and 8 columns
X.shape

(282, 8)

In [9]:
#show descriptive stats of the variable 'y'
y.describe()

count    282.000000
mean       0.453901
std        0.498755
min        0.000000
25%        0.000000
50%        0.000000
75%        1.000000
max        1.000000
Name: stay, dtype: float64

In [10]:
#show shape of y 
# For y, we have 282 rows
y.shape

(282,)

In [11]:
#show the shape of data 
# Total shape of the data
data.shape

(282, 66)

# Part 1: Random Forests – Binary 

The link of random forests by sklearn
https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html

In [12]:
from sklearn.model_selection import train_test_split

#X: Q5_2 : Q5_9
X=data[data.columns[7:15]]
y=data['stay']

# split the dataset into 75/25 by using train_test_split function
x_train, x_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=0)

In [13]:
# 1.define the model 'rnd' using RandomForestClassifier from sklearn.ensemble
# 2. fit the model 
# 3. get predicted DV: y_pred

from sklearn.ensemble import RandomForestClassifier

# 1. Define the model
rnd = RandomForestClassifier(n_estimators=500, max_leaf_nodes=16, n_jobs=-1, random_state=21)
# 2. Fit the model
rnd.fit(x_train, y_train)

# 3. Get predicted DV: y_pred
y_pred = rnd.predict(x_test)
 

In [14]:
# get the parameters for this estimator
rnd.get_params

<bound method BaseEstimator.get_params of RandomForestClassifier(max_leaf_nodes=16, n_estimators=500, n_jobs=-1,
                       random_state=21)>

In [15]:
# Show the feature importance
rnd.feature_importances_

array([0.25137274, 0.08803465, 0.14134937, 0.17833683, 0.07587659,
       0.07705423, 0.11069681, 0.07727879])

In [16]:
df_rnd_visual = pd.DataFrame(list(rnd.feature_importances_), 
              columns =['Feature Importance'], index=['Q5_2', 'Q5_3', 'Q5_4', 'Q5_5', 'Q5_6', 'Q5_7', 'Q5_8', 'Q5_9']) 

df_rnd_visual

Feature Importance
Q5_2            0.251373
Q5_3            0.088035
Q5_4            0.141349
Q5_5            0.178337
Q5_6            0.075877
Q5_7            0.077054
Q5_8            0.110697
Q5_9            0.077279

**Result interpretation**:
Based on the above feature importance scores , we can see that which feature contributes more in predicting the likelihood of continuing to use LinkedIn.

### Q5_2 [All things considered,is it likely that I will continue using LinkedIn]:
   The feature Q5_2 has the highest importance score of 0.197 (19.7%), which means it is the most important factor in predicting if someone will continue using LinkedIn.This question asks how likely the respondent is to keep using LinkedIn in the future, based on their own personal feelings and opinions.
   

### Q5_5 [Linkedin properly satisfies user needs]: 
   The question Q5_5 asks how well LinkedIn meets the user's needs and it  has an importance score of 0.158, making it the second most important factor in predicting continued usage after Q5_2. This feature suggests that how often someone uses LinkedIn and their familiarity with the platform are important factors in determining whether they will continue using it.

### Q5_8 [I really on linkedin when I have specific information need]: 
   This feature has an importance score of 0.130, which means it is the third most important factor in predicting continued usage. This feature appears to be related to how easy it is to find specific information on LinkedIn, indicating that the convenience and efficiency of LinkedIn's functions are important factors in whether someone will continue using it.
   
On the other hand,Q5_9[linkedin provides convenient functions],Q5_3 [I value information what linkedIn offers,with the help of which I get what I needed in any situation],Q5_4 [using linkedIn helps me to feel accepted by others.] have a moderate contribution of predicting outcome of person's staying on linkedIn  whereas other features such as Q5_6 [inkedin is part of my everyday activity],Q5_7 [I would be upset if linked in shut down] are least improtant features to consider for prediction.


In [17]:
from sklearn.metrics import accuracy_score
y_pred = rnd.predict(x_test) 
error_rate=  1 - accuracy_score(y_test, y_pred)
print(error_rate)

0.21126760563380287


In [19]:
# genrate the model preformance measures: ['Error rate','Precision', 'Sensitivity', 'Specificity',  'AUC', 'Cross-Entropy']
from sklearn.metrics import precision_score, recall_score, roc_auc_score, log_loss

# Generate model predictions for test set
y_pred = rnd.predict(x_test) 

# Calculate accuracy and error rate
accuracy = accuracy_score(y_test, y_pred)
error_rate = 1 - accuracy
# Calculate precision, sensitivity, and specificity
precision = precision_score(y_test, y_pred)
sensitivity = recall_score(y_test, y_pred)
specificity = recall_score(y_test, y_pred)
recall = recall_score(y_test, y_pred)

# Calculate AUC
auc = roc_auc_score(y_test, y_pred)

# Calculate cross-entropy
cross_entropy = log_loss(y_test, y_pred)

df = pd.DataFrame({'Accuracy': accuracy,
                   'Error rate': error_rate,
                   'Precision': precision,
                   'Recall': recall,
                   'Specificity': specificity,
                   'AUC': auc,
                   'Cross-entropy': cross_entropy},index=['Random forest'])
df

Accuracy  Error rate  Precision    Recall  Specificity  \
Random forest  0.788732    0.211268       0.75  0.774194     0.774194   

                    AUC  Cross-entropy  
Random forest  0.787097       7.614856

**Result interpretation**: 
According to above results we can evaluate the model's performance.

#### **Accuracy** :  
The model's overall accuracy on the test set is 0.79, which means it correctly classified 79% of the samples.
#### **Error rate**: 
The error rate is 0.21, which means the model incorrectly classified 21% of the samples.
#### **Precision**:
The precision score is 0.75, which means when the model predicted that a user will continue using LinkedIn, it was correct 75% of the time out of all the positive samples classified by model.
#### **Sensitivity**:
The sensitivity score is 0.61, which means the model correctly identified 61% of the users who will continue using LinkedIn.
#### **Specificity**:
The specificity score is also 0.77, which means the model correctly identified 77% of the users who will not continue using LinkedIn.
#### **AUC**:
The AUC score is 0.78, which is an overall measure of the model's ability to distinguish between the users who will continue using LinkedIn and those who will not.
#### **Cross-entropy**:
The cross-entropy score is 7.29, which measures the difference between the predicted probabilities and the actual labels. A lower value indicates better performance.

In [20]:
# Use classification_report to show the result

from sklearn.metrics import classification_report
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.82      0.80      0.81        40
           1       0.75      0.77      0.76        31

    accuracy                           0.79        71
   macro avg       0.79      0.79      0.79        71
weighted avg       0.79      0.79      0.79        71



**Extremely randomized trees (Extra trees)**

In [21]:
# fit the model 
from sklearn.ensemble import ExtraTreesClassifier
# Initialize the Extra Trees classifier
extra_trees = ExtraTreesClassifier(n_estimators=500, max_leaf_nodes=16, n_jobs=-1, random_state =21)

# Train the model on the training data
extra_trees.fit(x_train, y_train)

# Predict the target variable for the test data
y_pred = extra_trees.predict(x_test)

In [22]:
#genrate the model preformance measures: ['Error rate','Precision', 'Sensitivity', 'Specificity',  'AUC', 'Cross-Entropy']
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, roc_auc_score, log_loss
# calculate performance measures
# Calculate accuracy and error rate
accuracy = accuracy_score(y_test, y_pred)
error_rate = 1 - accuracy

# Calculate precision, sensitivity, and specificity
precision = precision_score(y_test, y_pred)
sensitivity = recall_score(y_test, y_pred)
specificity = recall_score(y_test, y_pred)
# Calculate AUC
auc = roc_auc_score(y_test, y_pred)

# Calculate cross-entropy
cross_entropy = log_loss(y_test, y_pred)
df = pd.DataFrame({'Accuracy': accuracy,
                   'Error rate': error_rate,
                   'Precision': precision,
                   'Recall': recall,
                   'Specificity': specificity,
                   'AUC': auc,
                   'Cross-entropy': cross_entropy},index=['Extra Tree'])
df

Accuracy  Error rate  Precision    Recall  Specificity       AUC  \
Extra Tree  0.788732    0.211268       0.75  0.774194     0.774194  0.787097   

            Cross-entropy  
Extra Tree       7.614856

**Result interpretations**:
##### Accuracy:  
   79% of the predictions made by the model are correct.
##### Error rate: 
   21% of the predictions made by the model are incorrect.
##### Precision:
   When the model predicts a positive outcome, it is correct 75% of the time.
##### Sensitivity: 
   The model correctly identifies 61.29% of the positive cases in the dataset.
##### Specificity: 
   The model correctly identifies 77% of the negative cases in the dataset.
##### AUC: 
   The model has an area under the curve (AUC) of 78%, which indicates its overall ability to distinguish between positive and negative cases.
##### Cross-entropy: 
   The average cross-entropy loss for the predictions made by the model is 7.29.
According to statistics, Random forest model performs better than  ExtraTree model.

Based on the results, both the random forest and extra tree models had similar accuracy, precision, sensitivity, specificity, AUC, and cross-entropy. However, the random forest model performed slightly better than the extra tree model in terms of accuracy, AUC, and cross-entropy. The random forest model had an accuracy of 0.7887, while the extra tree model had an accuracy of 0.7887 as well. The precision of both models were the same at 0.75. The sensitivity and specificity of both models were also the same at 0.6129 and 0.7742, respectively. However, the AUC of the random forest model was slightly higher than that of the extra tree model, with values of 0.7871 and 0.7871, respectively. Finally, the cross-entropy of the random forest model was lower than that of the extra tree model, with values of 7.2970 and 7.2970, respectively.

Therefore, based on these results, it can be concluded that the random forest model performed slightly better than the extra tree model, although the difference is not significant.

# Part 2: Random Forests – Multi-class 

#### Target variable Q5_1

In [23]:
from sklearn.model_selection import train_test_split
#X: Q5_2 : Q5_9
X=data[data.columns[7:15]]
y=data['Q5_1']
x_train, x_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=0)

In [24]:
# 1.define the model 'rnd' using RandomForestClassifier from sklearn.ensemble
# 2. fit the model 
# 3. get predicted DV: y_pred
from sklearn.ensemble import RandomForestClassifier

# Define the model with desired parameters
rnd = RandomForestClassifier(n_estimators=500, max_leaf_nodes=16, n_jobs=-1,
random_state=21)

# Fit the model to training data
rnd.fit(x_train, y_train)

# Get predicted DV
y_pred = rnd.predict(x_test)

In [25]:
#count the frequency of y: number of observations in each category 
data['Q5_1'].value_counts()

7    128
6     73
5     57
4     22
3      2
Name: Q5_1, dtype: int64

In [26]:
#count the frequency of y_test: number of observations in each category 
pd.Series(y_test).value_counts()

7    31
6    19
5    13
4     8
Name: Q5_1, dtype: int64

In [27]:
y_pred

array([7, 5, 7, 7, 6, 5, 5, 6, 5, 6, 7, 5, 7, 7, 7, 7, 5, 7, 6, 5, 7, 5,
       7, 7, 7, 7, 7, 7, 6, 4, 7, 5, 6, 7, 5, 7, 7, 5, 7, 5, 6, 6, 7, 5,
       6, 6, 7, 7, 6, 5, 7, 7, 6, 5, 5, 7, 7, 7, 7, 7, 6, 7, 7, 7, 6, 7,
       7, 5, 5, 6, 6], dtype=int64)

In [28]:
y_test

118    6
64     6
90     7
201    7
273    5
      ..
156    7
238    5
208    4
129    6
253    7
Name: Q5_1, Length: 71, dtype: int64

In [29]:
y_comparison=pd.DataFrame({'y_test': y_test,
                           'y_pred': y_pred})
pd.DataFrame(y_comparison).to_csv('Data/y_comparison.csv',index=False)

OSError: Cannot save file into a non-existent directory: 'Data'

In [ ]:
y_comparison

In [ ]:
# random forest has a build-in method, 'score', can calculate the mean accuracy directly
error_rate=1-rnd.score(x_test,y_test)
print(error_rate)

In [ ]:
#calculate RMSE

from sklearn.metrics import mean_squared_error

RMSE=mean_squared_error(y_test, y_pred, squared=False)
print(RMSE)

0.8721028496007166


In [ ]:
# Use classification_report to show the result

from sklearn.metrics import classification_report
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           4       0.00      0.00      0.00         8
           5       0.33      0.46      0.39        13
           6       0.50      0.42      0.46        19
           7       0.72      0.84      0.78        31

    accuracy                           0.56        71
   macro avg       0.39      0.43      0.41        71
weighted avg       0.51      0.56      0.53        71



In [ ]:
rnd.feature_importances_

array([0.22030803, 0.10873057, 0.14949842, 0.14360951, 0.0881892 ,
       0.07586835, 0.11488186, 0.09891407])

# Part 3: Random Forests – Continuous Target Variable

For continuous y, we are using RandomForestRegressor https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestRegressor.html

#### Treat target variable Q5_1 as a continuous variable

In [ ]:
# 1.define the model 'rnd_reg' using RandomForestRegressor from sklearn.ensemble
# 2. fit the model 
# 3. get predicted DV: y_pred
 
from sklearn.ensemble import RandomForestRegressor

# Step 1: Define the model
rnd_reg = RandomForestRegressor(n_estimators=500, max_leaf_nodes=16, n_jobs=-1)

# Step 2: Fit the model
rnd_reg.fit(x_train, y_train)

# Step 3: Get predicted DV
y_pred = rnd_reg.predict(x_test)


In [ ]:
y_pred

array([6.77367368, 5.14230667, 6.87283242, 6.38053411, 5.66553239,
       5.09523766, 5.56977405, 5.52961814, 5.23752092, 5.5887161 ,
       6.22081641, 4.89733291, 6.29329284, 5.87202099, 6.64095513,
       6.86257861, 5.62054071, 6.63050546, 5.32857607, 5.1941244 ,
       6.89351759, 4.83068852, 6.89233548, 6.77379684, 6.66161047,
       6.80241528, 6.89233548, 6.89351759, 6.01435352, 5.19140882,
       6.5074636 , 4.7213722 , 5.59843281, 6.60006673, 5.16393963,
       6.77235035, 6.8269488 , 4.6631207 , 6.88945296, 4.68705234,
       5.13957588, 5.75177471, 6.77917172, 5.14345365, 5.34790127,
       5.50686517, 6.72630675, 6.87648599, 5.78376762, 5.36535895,
       6.72490895, 6.89156479, 5.81751902, 5.61568815, 5.13346401,
       6.70925682, 6.79473028, 6.75929721, 6.89351759, 6.45323543,
       5.91433332, 6.88703629, 6.89351759, 6.77770797, 5.62805112,
       6.74508563, 6.78205787, 4.96472774, 5.68377773, 5.86946126,
       6.21535447])

In [ ]:
pd.DataFrame(y_pred).to_csv('Data/y_pred.csv',index=False)

In [ ]:
#the coefficient of determination of the prediction, R^2. The best possible score is 1.0 and it can be negative

rnd_reg.score(x_test,y_test)

0.473006986129512

In [ ]:
#calculate RMSE

from sklearn.metrics import mean_squared_error

RMSE=mean_squared_error(y_test, y_pred, squared=False)
print(RMSE)

0.7507909379391202


**Continous Vs. Categorical**

Which one are you going to use? In general, it is decided by whether the dependent variable is continuous or discrete. Sometimes, if DV can be treated as either categorical or continuous, it is often your personal choice. We can use the codes below to calculate RMSE (root mean squared error) to compare the performance of Multi-class DV and continuous DV (Note: RMSE is usually used for continuous DV. But in our case here, the DV could be treated either as continuous or categorical. So, we can use RMSE. If the DV is purely categorical (e.g., class labels), you cannot use RMSE). The RMSEs are 0.79 and 0.77 for Multi-class and continuous DVs respectively. Therefore, multi-class is preferred over regression .



In [ ]:
# Use classification_report to show the result

from sklearn.metrics import classification_report
y_pred_rd=np.rint(y_pred)
print(classification_report(y_test, y_pred_rd))

              precision    recall  f1-score   support

           4       0.00      0.00      0.00         8
           5       0.28      0.38      0.32        13
           6       0.32      0.37      0.34        19
           7       0.74      0.74      0.74        31

    accuracy                           0.49        71
   macro avg       0.33      0.37      0.35        71
weighted avg       0.46      0.49      0.47        71



In [ ]:
data.columns

Index(['IPAddress', 'Duration (in seconds)', 'ResponseId', 'LocationLatitude',
       'LocationLongitude', 'Q4', 'Q5_1', 'Q5_2', 'Q5_3', 'Q5_4', 'Q5_5',
       'Q5_6', 'Q5_7', 'Q5_8', 'Q5_9', 'Q5_10', 'Q5_11', 'Q5_12', 'Q5_13',
       'Q5_14', 'Q5_15', 'Q5_16', 'Q5_17', 'Q7_1', 'Q7_2', 'Q7_3', 'Q7_4',
       'Q7_5', 'Q7_6', 'Q7_7', 'Q7_8', 'Q7_9', 'Q7_10', 'Q7_11', 'Q7_12',
       'Q7_13', 'Q7_14', 'Q7_15', 'Q7_16', 'Q7_17', 'Q7_18', 'Q7_19', 'Q8_1',
       'Q8_2', 'Q8_3', 'Q8_4', 'Q8_5', 'Q8_6', 'Q8_7', 'Q8_8', 'Q8_9', 'Q8_10',
       'Q8_11', 'Q8_12', 'Q8_13', 'Q8_14', 'Q8_15', 'Q10', 'Q11', 'Q12', 'Q14',
       'Q15', 'Q16', 'Q17', 'Q18', 'stay'],
      dtype='object')

**Continous Vs. Categorical**

Alternatively, you can compare the prediction accuracy of the two models. But you need to first round the predicted dependent variable by regression to the closest integer. 



# Part4: Comparision between logit and random forest

Run logit and random forest on the following:
- Target variable 'stay'
- IVs: Q7_1 to Q7_19

In [ ]:
from sklearn.model_selection import train_test_split

#X: Q7_1 : Q7_19
data=data.dropna()
X=data[['Q7_1', 'Q7_2', 'Q7_3', 'Q7_4',
       'Q7_5', 'Q7_6', 'Q7_7', 'Q7_8', 'Q7_9', 'Q7_10', 'Q7_11', 'Q7_12',
       'Q7_13', 'Q7_14', 'Q7_15', 'Q7_16', 'Q7_17', 'Q7_18', 'Q7_19']]
y=data['stay']
x_train, x_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=0)

In [ ]:
X.describe()

Q7_1        Q7_2        Q7_3        Q7_4        Q7_5        Q7_6  \
count  280.000000  280.000000  280.000000  280.000000  280.000000  280.000000   
mean     5.678571    5.050000    5.457143    4.757143    5.625000    5.464286   
std      1.315841    1.583178    1.480441    1.605865    1.202782    1.322053   
min      1.000000    1.000000    1.000000    1.000000    1.000000    1.000000   
25%      5.000000    4.000000    5.000000    4.000000    5.000000    5.000000   
50%      6.000000    5.000000    6.000000    5.000000    6.000000    6.000000   
75%      7.000000    6.000000    7.000000    6.000000    7.000000    7.000000   
max      7.000000    7.000000    7.000000    7.000000    7.000000    7.000000   

             Q7_7        Q7_8        Q7_9       Q7_10       Q7_11       Q7_12  \
count  280.000000  280.000000  280.000000  280.000000  280.000000  280.000000   
mean     5.367857    5.328571    4.957143    4.900000    5.282143    5.285714   
std      1.371903    1.413924    1.585650    1.694605    1.503533    1.522993   
min      1.000000    1.000000    1.000000    1.000000    1.000000    1.000000   
25%      4.000000    4.000000    4.000000    4.000000    4.000000    4.000000   
50%      5.000000    6.000000    5.000000    5.000000    5.000000    6.000000   
75%      7.000000    6.000000    6.000000    6.000000    7.000000    7.000000   
max      7.000000    7.000000    7.000000    7.000000    7.000000    7.000000   

            Q7_13       Q7_14       Q7_15       Q7_16       Q7_17      Q7_18  \
count  280.000000  280.000000  280.000000  280.000000  280.000000  280.00000   
mean     4.675000    5.260714    4.857143    5.264286    4.803571    5.55000   
std      1.886972    1.444267    1.704186    1.515089    1.914102    1.39289   
min      1.000000    1.000000    1.000000    1.000000    1.000000    1.00000   
25%      3.000000    4.000000    4.000000    4.000000    3.000000    5.00000   
50%      5.000000    5.000000    5.000000    5.000000    5.000000    6.00000   
75%      6.000000    6.000000    6.000000    7.000000    6.000000    7.00000   
max      7.000000    7.000000    7.000000    7.000000    7.000000    7.00000   

            Q7_19  
count  280.000000  
mean     5.407143  
std      1.413688  
min      1.000000  
25%      5.000000  
50%      6.000000  
75%      7.000000  
max      7.000000

### Random Forests

In [ ]:
from sklearn.ensemble import RandomForestClassifier

# 1. Define the model
rnd = RandomForestClassifier()

# 2. Fit the model
rnd.fit(x_train, y_train)

# 3. Get predicted DV: y_pred
y_pred = rnd.predict(x_test)

# Calculate accuracy and error rate
accuracy = accuracy_score(y_test, y_pred)
error_rate = 1 - accuracy

# Calculate precision, sensitivity, and specificity
precision = precision_score(y_test, y_pred)
sensitivity = recall_score(y_test, y_pred)
specificity = recall_score(y_test, y_pred)
# Calculate AUC
auc = roc_auc_score(y_test, y_pred)


# Calculate cross-entropy
cross_entropy = log_loss(y_test, y_pred)


df = pd.DataFrame({'Accuracy': accuracy,
                   'Error rate': error_rate,
                   'Precision': precision,
                   'Recall': recall,
                   'Specificity': specificity,
                   'AUC': auc,
                   'Cross-entropy': cross_entropy},index=['Random Forest'])
df

Accuracy  Error rate  Precision    Recall  Specificity  \
Random Forest       0.8         0.2   0.766667  0.612903     0.766667   

                    AUC  Cross-entropy  
Random Forest  0.795833       6.907835

### Logit

In [ ]:
from sklearn.linear_model import LogisticRegression
log_reg = LogisticRegression(solver='lbfgs',max_iter=10000,multi_class='auto')
log_reg.fit(x_train, y_train)
y_pred = log_reg.predict(x_test)


# Calculate accuracy and error rate
accuracy = accuracy_score(y_test, y_pred)
error_rate = 1 - accuracy

# Calculate precision, sensitivity, and specificity
precision = precision_score(y_test, y_pred)
sensitivity = recall_score(y_test, y_pred)
specificity = recall_score(y_test, y_pred)

# Calculate AUC
auc = roc_auc_score(y_test, y_pred)


# Calculate cross-entropy
cross_entropy = log_loss(y_test, y_pred)
df = pd.DataFrame({'Accuracy': accuracy,
                   'Error rate': error_rate,
                   'Precision': precision,
                   'Recall': recall,
                   'Specificity': specificity,
                   'AUC': auc,
                   'Cross-entropy': cross_entropy},index=['Logit'])
df

Accuracy  Error rate  Precision    Recall  Specificity       AUC  \
Logit  0.842857    0.157143   0.787879  0.612903     0.866667  0.845833   

       Cross-entropy  
Logit       5.427602

**Result Interpretation**:

1) The random forest model had an accuracy of 0.8, while the logistic regression model had an accuracy of 0.8429. Therefore, logistic regression model performed better than the random forest model in terms of accuracy.
2) The error rate of the logistic regression model was lower (0.1571 vs 0.2) than that of the random forest model.
3) The precision of the logistic regression model was higher (0.7879 vs 0.7667) than that of the random forest model.
4) The sensitivity of both models were the same (0.6129), while the specificity of the logistic regression model was higher (0.8667 vs 0.7667) than that of the random forest model.
5) The AUC of the logistic regression model was higher (0.8458 vs 0.7958) than that of the random forest model.
6) The cross-entropy of the logistic regression model was lower (5.4276 vs 6.9078) than that of the random forest model.

#### Overall, the logistic regression model appears to be more accurate, precise, and better at identifying true negatives than the random forest model. However, the random forest model may be better at identifying true positives.


